src/poprox_recommender/components/diversifiers/locality_calibration.py

```diff
@@ -25,13 +47,234 @@ def __call__(self, candidate_articles: ArticleSet, interest_profile: InterestPro
+            for topic, click_count in interest_profile.click_topic_counts.items():
+                topic_preferences[topic] += click_count
+
+        normalized_topic_prefs = normalized_category_count(topic_preferences)
+        return normalized_topic_prefs
+
+
+###################### text generation part
+
+client = OpenAI(
+    api_key="PUT KEY HERE",
+)
+
+
+def gpt_generate(system_prompt, content_prompt):
+    message = [{"role": "system", "content": system_prompt}, {"role": "user", "content": content_prompt}]
+    temperature = 0.2
+    max_tokens = 512
+    frequency_penalty = 0.0
+
+    chat_completion = client.chat.completions.create(
+        messages=message,
+        temperature=temperature,
+        max_tokens=max_tokens,
+        frequency_penalty=frequency_penalty,
+        model="gpt-4o-mini",
+    )
+    return chat_completion.choices[0].message.content
+
+
+def news_factor(headline, slug):
+    news = f"headline: {headline}\nslug: {slug}"
+    system_prompt = (
+        "Based on the given news information, summarize what topics that the news is related to."
+        "Each news article is related to 1-3 topics and each topic shuold not exceed five words."
+    )
+    return gpt_generate(system_prompt, news)
+
+
+# TODO: check backward or forward for past k articles
+def user_interest_generate(past_articles: Article, past_k: int):
+    past_article_infor = []
+    past_count = 0
+    for article in past_articles.articles:
+        past_count += 1
+        if past_count == past_k:
+            break
+        temp_article = {}
+        temp_article["headline"] = article.headline
+        temp_article["subhead"] = article.subhead
+        temp_article["categories"] = set([mention.entity.name for mention in article.mentions])
+        temp_article["topics"] = news_factor(article.headline, article.subhead)
+        past_article_infor.append(temp_article)
+
+    system_prompt = (
+        "You are asked to describe user interest based on his/her browsed news list."
+        " User interest includes the news [categories] and news [topics]"
+        " (under each [category] that users are interested in."
+    )
+
+    return gpt_generate(system_prompt, f"{past_article_infor}")
+
+
+def generate_narrative(news_list, user_interests):
+    system_prompt = (
+        "You are a personalized text generator."
+        " First, i will provide you with a news list that"
+        " includes both the [main news] and [related news]."
+        " Second, I will provide you with user interests,"
+        " including the [categories] of news that the user is interested in."
+        " Based on the input news list and user interests,"
+        " you are required to generate a personalized news summary centered around the [main news]."
+    )
+
+    input_prompt = "News List: \n" + f"{news_list}" + "User Interest: \n" + user_interests
+    return gpt_generate(system_prompt, input_prompt)
+
+
+model = SentenceTransformer("all-MiniLM-L6-v2")
```

Comment:
Makes sense that you'd need to embed the subheads separately here since the article embedder only does the headlines right now. I'd probably use the same language model for this that the NRMS model is based on, which I think is `distilbert-base-uncased`

Also worth thinking about whether the article embedder should be configurable to specify which article text fields (headline, subhead, body text) to embed. You can do it here but this probably isn't the last time we're going to run into this, so might make sense to move it upstream and make it re-usable.

Reply:
I’m thinking, for handling headlines, should we use the original distilbert transformer, or can we easily extract embeddings from our fine-tuned distilbert on the MIND data? I agree that making this part more flexible for selecting configurations and text fields would be beneficial.

Reply:
It seems like it should be possible to use the fine-tuned one by loading the fine-tuned weights into a plain distilbert model